### PR TITLE
DO NOT MERGE - PR1 Triggers with final API v4

### PR DESCRIPTION
**Pull request by thoralf-gutierrez**

### CODE REVIEW

File: airflow/models.py [hunk 1]
```diff
@@ -47,12 +47,12 @@
...
-    FERNET = Fernet(configuration.get('core', 'FERNET_KEY').encode('utf-8'))
-    ENCRYPTION_ON = True
-except:
-    pass
+# try:
+#    from cryptography.fernet import Fernet
+#    FERNET = Fernet(configuration.get('core', 'FERNET_KEY').encode('utf-8'))
+#    ENCRYPTION_ON = True
+# except:
+#    pass
```
Comment (sv3ndk) [inline airflow/models.py:56]: SI tu rebase au dessus de ceci: https://github.com/RealImpactAnalytics/airflow/tree/add_frenet_key_to_test_config tu devrais pouvoir remettre ça en place (en supposant qu'il finisse par merge la branch ci-dessus)

File: airflow/models.py [hunk 2]
```diff
@@ -1652,13 +1645,48 @@ def resolve_template_files(self):
...
+        depends on that were added through set_downstream and set_upstream
+        only"""
+        return self._upstream_tuples
+
+    @property
+    def upstream_tuples_extra(self):
+        """@property: tuples of all the (task, trigger) on which this task
+        depends on that were added through depends_on_past or
+        wait_for_downstream only"""
+        tuples = self._upstream_tuples_extra
```
Comment (sv3ndk) [inline airflow/models.py]: maybe we could move the 

```
 if self.depends_on_past:
            trigger = Trigger(self, state=State.SUCCESS, past_executions=1)
            self._upstream_tuples_extra.append((self, trigger))
            self._downstream_tuples_extra.append((self, trigger))
```

here such that we would no longer need the `self._upstream_tuples_extra` but directly build the `tuples` list on the fly here ?

File: airflow/models.py [hunk 3]
```diff
@@ -1652,13 +1645,48 @@ def resolve_template_files(self):
...
+
+    @property
+    def upstream_triggers(self):
+        """@property: all the triggers on which this task depends on that
+        were added through set_downstream and set_upstream only"""
+        return [trigger for (_, trigger) in self.upstream_tuples]
+
+    @property
+    def upstream_triggers_extra(self):
+        """@property: all the triggers on which this task depends on that
```
Comment (sv3ndk) [inline airflow/models.py:1655]: cf remark below

File: airflow/models.py (hunk 2 above)
Comment (thoralf-gutierrez) [inline airflow/models.py]: Good idea ! We also get rid of `self._downstream_tuples_extra`

File: airflow/models.py [hunk 4]
```diff
@@ -711,131 +711,65 @@ def is_runnable(self, flag_upstream_failed=False):
...
-        """
-        TI = TaskInstance
-        TR = TriggerRule
+        # These dependencies are of the type ALL by default
+        triggered_extra = [x for (x, _) in evaluations_extra]
+        if all(triggered_extra):
+            triggered_extra_final = True
+        else:
+            triggered_extra_final = False
+
```
Comment (sv3ndk) [inline airflow/models.py:738]:
```
triggered_extra_final =  all(triggered_extra)
```

**Pull request closed without merging**